pandas/core/indexes/base.py

```diff
@@ -2246,6 +2246,25 @@ def _get_reconciled_name_object(self, other):
             return self._shallow_copy(name=name)
         return self
 
+    def _union_incompatible_dtypes(self, other):
+        """
+        Casts this and other index to object dtype to allow the formation
+        of a union between incompatible types.
+        """
+        this = self.astype(object)
+        # call Index for when `other` is list-like
+        other = Index(other).astype(object)
+        return Index.union(this, other).astype(object)
```

Comment:
do we need the .astype?

Reply:
Doing this causes several failures in incompatible dtype union tests, that seem to arise from union results that are automatically cast to a specific dtype after forming a union between 2 object dtype indexes.

Reply:
add .astype(object, copy=False) here (as it *already* maybe object)